### ISSUE TITLE
Error creating native image for Micronaut and AWS Dynamo DB without `--report-unsupported-elements-at-runtime`

### ISSUE THREAD

**Issue by ilopmar** [oracle/graal#2758]
**Describe the issue**

To create a Micronaut native image with AWS Dynamo DB support is necessary to use the discouraged flag `--report-unsupported-elements-at-runtime`.

**Steps to reproduce the issue**

- `git clone https://github.com/micronaut-core-issues-test-app/issue-aws-454-graalvm`
- `cd issue-aws-454-graalvm`
- Use graalvm from master branch
- `./build-native-image.sh`

The application fails.

Add flag `--report-unsupported-elements-at-runtime` and generate the native image again. Then the native image is generated correctly.

**Describe GraalVM and your environment:**
 - GraalVM version: **20.1.0** and master commit **e666b90d94244662a51abded90c0f861226d2dd7**
 - JDK major version: **8**
 - OS: **Linux Mint 20**
 - Architecture: **AMD64**

**More details**
Consider adding the `--native-image-info` and  `--verbose` flags when building your native image and paste output below.

```
Executing [
/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/bin/java \
-XX:+UseParallelGC \
-XX:+UnlockExperimentalVMOptions \
-XX:+EnableJVMCI \
-Dtruffle.TrustAllTruffleRuntimeProviders=true \
-Dtruffle.TruffleRuntime=com.oracle.truffle.api.impl.DefaultTruffleRuntime \
-Dgraalvm.ForcePolyglotInvalid=true \
-Dgraalvm.locatorDisabled=true \
-d64 \
-XX:-UseJVMCIClassLoader \
-XX:+UseJVMCINativeLibrary \
-Xss10m \
-Xms1g \
-Xmx14g \
-Duser.country=US \
-Duser.language=en \
-Djava.awt.headless=true \
-Dorg.graalvm.version=20.3.0-dev \
-Dorg.graalvm.config= \
-Dcom.oracle.graalvm.isaot=true \
-Djava.system.class.loader=com.oracle.svm.hosted.NativeImageSystemClassLoader \
-Xshare:off \
-Djvmci.class.path.append=/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/jvmci/graal.jar \
-javaagent:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/svm/builder/svm.jar \
-Djdk.internal.lambda.disableEagerInitialization=true \
-Djdk.internal.lambda.eagerlyInitialize=false \
-Djava.lang.invoke.InnerClassLambdaMetafactory.initializeLambdas=false \
-Xbootclasspath/a:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/boot/graal-sdk.jar \
-cp \
/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/svm/builder/svm.jar:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/svm/builder/objectfile.jar:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/svm/builder/pointsto.jar:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/svm/builder/svm-llvm.jar:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/svm/builder/llvm-platform-specific-shadowed.jar:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/svm/builder/javacpp-shadowed.jar:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/svm/builder/llvm-wrapper-shadowed.jar:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/jvmci/graal-truffle-jfr-impl.jar:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/jvmci/jvmci-hotspot.jar:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/jvmci/graal-management.jar:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/jvmci/jvmci-api.jar:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/jvmci/graal.jar:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/resources.jar \
com.oracle.svm.hosted.NativeImageGeneratorRunner \
-watchpid \
83693 \
-imagecp \
/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/boot/graal-sdk.jar:/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/svm/library-support.jar:/home/ivan/workspaces/micronaut-apps/issue-aws-454-graal/build/libs/demo-0.1-all.jar \
-H:Path=/home/ivan/workspaces/micronaut-apps/issue-aws-454-graal \
-H:FallbackThreshold=0 \
-H:ClassInitialization=io.netty.util.AbstractReferenceCounted:run_time,io.netty.util.concurrent.GlobalEventExecutor:run_time,io.netty.util.concurrent.ImmediateEventExecutor:run_time,io.netty.util.concurrent.ScheduledFutureTask:run_time,io.netty.util.internal.ThreadLocalRandom:run_time \
-H:ClassInitialization=io.netty.buffer.PooledByteBufAllocator:run_time,io.netty.buffer.ByteBufAllocator:run_time,io.netty.buffer.ByteBufUtil:run_time,io.netty.buffer.AbstractReferenceCountedByteBuf:run_time \
-H:ReflectionConfigurationResources=META-INF/native-image/io.netty/transport/reflection-config.json \
-H:ClassInitialization=io.netty.handler.ssl.util.ThreadLocalInsecureRandom:run_time \
-H:ClassInitialization=io.netty.handler.codec.http.HttpObjectEncoder:run_time,io.netty.handler.codec.http.websocketx.WebSocket00FrameEncoder:run_time,io.netty.handler.codec.http.websocketx.extensions.compression.DeflateDecoder:run_time \
-H:ClassInitialization=io.netty:build_time \
-H:ClassInitialization=io.netty.handler.codec.http2.Http2CodecUtil:run_time,io.netty.handler.codec.http2.Http2ClientUpgradeCodec:run_time,io.netty.handler.codec.http2.Http2ConnectionHandler:run_time,io.netty.handler.codec.http2.DefaultHttp2FrameWriter:run_time \
-H:ReflectionConfigurationResources=META-INF/native-image/io.micronaut.core/cache-caffeiene-graal/reflection-config.json \
-H:IncludeResources=META-INF/http/mime.types \
-H:ClassInitialization=io.micronaut.http.bind.binders.ContinuationArgumentBinder:run_time \
-H:ClassInitialization=io.micronaut.function.aws.proxy:run_time \
-H:ClassInitialization=io.micronaut.function.aws.proxy.model.factory.MicronautAwsProxyResponseFactory:build_time \
-H:+AllowIncompleteClasspath \
-H:EnableURLProtocols=http,https \
-H:ReflectionConfigurationResources=META-INF/native-image/io.micronaut/http/reflection-config.json \
'-H:ClassInitialization=com.sun.jndi.dns.DnsClient:run_time,io.netty.handler.ssl.ConscryptAlpnSslEngine:run_time,io.netty.handler.ssl.JettyNpnSslEngine:run_time,io.netty.handler.ssl.ReferenceCountedOpenSslEngine:run_time,io.netty.handler.ssl.JdkNpnApplicationProtocolNegotiator:run_time,io.netty.handler.ssl.ReferenceCountedOpenSslServerContext:run_time,io.netty.handler.ssl.ReferenceCountedOpenSslClientContext:run_time,io.netty.handler.ssl.util.BouncyCastleSelfSignedCertGenerator:run_time,io.netty.handler.ssl.ReferenceCountedOpenSslContext:run_time,io.micronaut.buffer.netty.NettyByteBufferFactory:run_time,io.netty.handler.ssl.JettyAlpnSslEngine$ClientEngine:run_time,io.netty.handler.ssl.JettyAlpnSslEngine$ServerEngine:run_time,io.netty.handler.codec.http2.Http2CodecUtil:run_time,io.netty.handler.codec.http2.CleartextHttp2ServerUpgradeHandler:run_time,io.netty.handler.codec.http2.Http2ServerUpgradeCodec:run_time,io.micronaut.http.netty.channel.converters.EpollChannelOptionFactory:run_time,io.micronaut.http.netty.channel.converters.KQueueChannelOptionFactory:run_time,io.micronaut.http.bind.binders.ContinuationArgumentBinder$Companion:run_time,io.micronaut.http.bind.binders.ContinuationArgumentBinder:run_time' \
-H:ReflectionConfigurationResources=META-INF/native-image/io.micronaut/http-netty/reflection-config.json \
-H:ClassInitialization=io.micronaut.reactive.reactor.ReactorInstrumentation:run_time \
-H:ClassInitialization=ch.qos.logback:build_time,com.fasterxml.jackson:build_time,io.micronaut:build_time,io.reactivex:build_time,org.reactivestreams:build_time,org.slf4j:build_time,org.yaml.snakeyaml:build_time,javax:build_time \
-H:ClassInitialization=com.sun.org.apache.xerces.internal.util:build_time,com.sun.org.apache.xerces.internal.impl:build_time,jdk.xml.internal:build_time,com.sun.xml.internal.stream.util:build_time,com.sun.org.apache.xerces.internal.xni:build_time,com.sun.org.apache.xerces.internal.utils:build_time \
-H:ReflectionConfigurationResources=META-INF/native-image/io.micronaut/health/reflection-config.json \
-H:ReflectionConfigurationResources=META-INF/native-image/io.micronaut/function-aws-runtime/reflection-config.json \
-H:ReflectionConfigurationResources=META-INF/native-image/io.micronaut/function-aws-proxy/reflection-config.json \
-H:Name=demo \
-H:Class=io.micronaut.function.aws.runtime.MicronautLambdaRuntime \
-H:ReflectionConfigurationResources=META-INF/native-image/com.example/demo/reflection-config.json \
-H:ResourceConfigurationResources=META-INF/native-image/io.micronaut/http/resource-config.json \
-H:ResourceConfigurationResources=META-INF/native-image/com.example/demo/resource-config.json \
-H:+DumpTargetInfo \
-H:CLibraryPath=/home/ivan/workspaces/misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/svm/clibraries/linux-amd64 \

]
[demo:83710]    classlist:   6,436.80 ms,  2.32 GB
[demo:83710]        (cap):   1,054.04 ms,  2.32 GB
[demo:83710]        setup:   3,093.58 ms,  2.32 GB
# Building image for target platform: org.graalvm.nativeimage.Platform$LINUX_AMD64
# Using native toolchain:
#   Name: GNU project C and C++ compiler (gcc)
#   Vendor: linux
#   Version: 9.3.0
#   Target architecture: x86_64
#   Path: /usr/bin/gcc
# Using CLibrary: com.oracle.svm.core.posix.linux.libc.GLibC
[demo:83710]     (clinit):     982.68 ms,  3.52 GB
# Static libraries:
#   ../../../misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/svm/clibraries/linux-amd64/liblibchelper.a
#   ../../../misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/libnet.a
#   ../../../misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/svm/clibraries/linux-amd64/libffi.a
#   ../../../misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/libnio.a
#   ../../../misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/libjava.a
#   ../../../misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/libfdlibm.a
#   ../../../misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/libsunec.a
#   ../../../misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/libzip.a
#   ../../../misc/graal/graal/sdk/mxbuild/linux-amd64/GRAALVM_BCF68E8E09_JAVA8/graalvm-bcf68e8e09-java8-20.3.0-dev/jre/lib/svm/clibraries/linux-amd64/libjvm.a
# Other libraries: stdc++,pthread,dl,z,rt
[demo:83710]   (typeflow):  31,351.29 ms,  3.52 GB
[demo:83710]    (objects):  28,665.96 ms,  3.52 GB
[demo:83710]   (features):   4,638.56 ms,  3.52 GB
[demo:83710]     analysis:  68,777.78 ms,  3.52 GB
Error: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported method java.lang.ClassLoader.defineClass1(String, byte[], int, int, ProtectionDomain, String) is reachable
To diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The unsupported element is then reported at run time when it is accessed the first time.
Detailed message:
Trace: 
	at parsing java.lang.ClassLoader.defineClass(ClassLoader.java:756)
Call path from entry point to java.lang.ClassLoader.defineClass(String, byte[], int, int, ProtectionDomain): 
	at java.lang.ClassLoader.defineClass(ClassLoader.java:754)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:635)
	at com.oracle.svm.reflect.ClassLoader_defineClass_c6c343b4d6dc22ca64eb2d8503b13ac9c340dcb3_2496.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at software.amazon.awssdk.http.apache.internal.conn.ClientConnectionRequestFactory$Handler.invoke(ClientConnectionRequestFactory.java:83)
	at com.sun.proxy.$Proxy215.hashCode(Unknown Source)
	at java.util.HashMap.hash(HashMap.java:339)
	at java.util.HashMap.get(HashMap.java:557)
	at com.oracle.svm.jni.access.JNIReflectionDictionary.getClassObjectByName(JNIReflectionDictionary.java:129)
	at com.oracle.svm.jni.functions.JNIFunctions.FindClass(JNIFunctions.java:324)
	at com.oracle.svm.core.code.IsolateEnterStub.JNIFunctions_FindClass_3ec1032c6cb9443725d1e68194130533bfb04076(generated:0)

Error: Use -H:+ReportExceptionStackTraces to print stacktrace of underlying exception
Error: Image build request failed with exit status 1
com.oracle.svm.driver.NativeImage$NativeImageError: Image build request failed with exit status 1
	at com.oracle.svm.driver.NativeImage.showError(NativeImage.java:1571)
	at com.oracle.svm.driver.NativeImage.build(NativeImage.java:1321)
	at com.oracle.svm.driver.NativeImage.performBuild(NativeImage.java:1282)
	at com.oracle.svm.driver.NativeImage.main(NativeImage.java:1241)

```

**Comment by mcraj017** [comment]
@ilopmar In future, we will remove this flag, when native-image compatibly reaches a decent level

**Comment by ilopmar** [comment]
I see... I've always being told that the flag is discouraged, that's why every time I need to add it to be able to create the native-image, I create here an issue. Should I stop doing that?

**Comment by mcraj017** [comment]
@ilopmar you are always welcome to create an issue, when ever you use the flag.
if that answer your question, please let me know, if I can the issue?

**Comment by ilopmar** [comment]
This answer my question but I think this issue should only be closed once it's fixed. I mean, if you close it without a fix and in the next release (for example) you remove the flag, then there won't be a way to generate a native image for Micronaut + DynamoDB.

**Comment by cstancu** [comment]
Fixed by f4f1fc8ba8d2044df680ebe40bf2313ca2c62399. `ClassLoader.defineClass()` was automatically registered for reflection, but it should have just been `@Delete`ed and `@Delete`ed methods should be ignored by the reflection plugin. @ilopmar I think this fix should also apply to other similar open issues.
